File: apps/api/app/schemas/ai.py
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, Field


class GeneralAdviceRequest(BaseModel):
    question: str = Field(min_length=1, max_length=500)


class GeneralAdviceContextResponse(BaseModel):
    currency_code: str
    currency_symbol: str
    month_label: str
    current_month_income: Decimal
    current_month_expense: Decimal
    current_month_net: Decimal
    active_goal_count: int
    comfortable_monthly_savings: Decimal
    behavior_label: str
    behavior_score: int
    top_spending_category: str | None = None
    savings_rate: Decimal


class GeneralAdviceResponse(BaseModel):
    guidance: str
    provider: str
    model_name: str
    context: GeneralAdviceContextResponse


class PurchaseCheckRequest(BaseModel):
    planned_amount: Decimal = Field(gt=0)
    item_name: str = Field(min_length=1, max_length=120)
    question: str = Field(min_length=1, max_length=500)
    category_id: UUID | None = None


class PurchaseCheckContextResponse(BaseModel):
    currency_code: str
    currency_symbol: str
    month_label: str
    planned_amount: Decimal
    item_name: str
    category_name: str | None = None
    current_category_spend: Decimal | None = None
    category_budget_limit: Decimal | None = None
    current_month_income: Decimal
    current_month_expense: Decimal
    current_month_net: Decimal
    top_spending_category: str | None = None
    active_goal_count: int
    total_goal_monthly_required: Decimal
    comfortable_monthly_savings: Decimal
    affordability_ratio: Decimal
    verdict: str
    suggested_action: str


class PurchaseCheckResponse(BaseModel):
    verdict: str
    affordability_score: int
    guidance: str
    provider: str
    model_name: str
    context: PurchaseCheckContextResponse


class SavingsAdviceRequest(BaseModel):
    question: str = Field(min_length=1, max_length=500)
    goal_id: UUID | None = None


class SavingsAdviceAllocationResponse(BaseModel):
    goal_id: UUID
    name: str
    recommended_monthly_contribution: Decimal
    required_monthly_contribution: Decimal
    pace_status: str


class SavingsAdviceContextResponse(BaseModel):
    currency_code: str
    currency_symbol: str
    month_label: str
    current_month_income: Decimal
    current_month_expense: Decimal
    current_month_net: Decimal
    active_goal_count: int
    comfortable_monthly_savings: Decimal
    total_goal_monthly_required: Decimal
    overall_goal_progress: Decimal
    can_fund_all_goals_on_time: bool
    recommendation_text: str
    focus_goal_name: str | None = None
    focus_goal_progress_percentage: Decimal | None = None
    focus_goal_monthly_required: Decimal | None = None
    focus_goal_pace_status: str | None = None
    allocations: list[SavingsAdviceAllocationResponse]


class SavingsAdviceResponse(BaseModel):
    guidance: str
    provider: str
    model_name: str
    context: SavingsAdviceContextResponse


class SpendingSummaryRequest(BaseModel):
    question: str = Field(min_length=1, max_length=500)
    months: int = Field(default=4, ge=2, le=12)


class SpendingSummaryCategoryResponse(BaseModel):
    name: str
    total_amount: Decimal
    percentage: Decimal
    trend_direction: str


class SpendingSummaryContextResponse(BaseModel):
    currency_code: str
    currency_symbol: str
    period_label: str
    total_spent: Decimal
    behavior_label: str
    behavior_score: int
    planned_buys: int
    impulse_buys: int
    overspent_days: int
    strongest_insight_title: str | None = None
    strongest_insight_description: str | None = None
    top_categories: list[SpendingSummaryCategoryResponse]


class SpendingSummaryResponse(BaseModel):
    guidance: str
    provider: str
    model_name: str
    context: SpendingSummaryContextResponse


class ReportSummaryRequest(BaseModel):
    question: str = Field(min_length=1, max_length=500)
    months: int = Field(default=4, ge=2, le=12)


class ReportSummaryTransactionResponse(BaseModel):
    title: str
    amount: Decimal
    transaction_date: str
    category_name: str | None = None


class AIReportSummaryContextResponse(BaseModel):
    currency_code: str
    currency_symbol: str
    period_label: str
    total_income: Decimal
    total_expense: Decimal
    net_saved: Decimal
    transaction_count: int
    savings_rate: Decimal
    savings_rate_delta: Decimal | None = None
    top_category_name: str | None = None
    top_category_total_amount: Decimal | None = None
    largest_transactions: list[ReportSummaryTransactionResponse]


class AIReportSummaryResponse(BaseModel):
    guidance: str
    provider: str
    model_name: str
    context: AIReportSummaryContextResponse

File: apps/mobile/lib/api/ai.ts
import { apiRequest } from '@/lib/api/client';

export type GeneralAdvicePayload = {
  question: string;
};

export type GeneralAdviceContext = {
  currency_code: string;
  currency_symbol: string;
  month_label: string;
  current_month_income: string;
  current_month_expense: string;
  current_month_net: string;
  active_goal_count: number;
  comfortable_monthly_savings: string;
  behavior_label: string;
  behavior_score: number;
  top_spending_category: string | null;
  savings_rate: string;
};

export type GeneralAdviceResponse = {
  guidance: string;
  provider: string;
  model_name: string;
  context: GeneralAdviceContext;
};

export async function generalChat(accessToken: string, payload: GeneralAdvicePayload) {
  return apiRequest<GeneralAdviceResponse>('/ai/chat', {
    accessToken,
    body: payload,
    method: 'POST',
  });
}

export type PurchaseCheckPayload = {
  planned_amount: number;
  item_name: string;
  question: string;
  category_id?: string | null;
};

export type PurchaseCheckContext = {
  currency_code: string;
  currency_symbol: string;
  month_label: string;
  planned_amount: string;
  item_name: string;
  category_name: string | null;
  current_category_spend: string | null;
  category_budget_limit: string | null;
  current_month_income: string;
  current_month_expense: string;
  current_month_net: string;
  top_spending_category: string | null;
  active_goal_count: number;
  total_goal_monthly_required: string;
  comfortable_monthly_savings: string;
  affordability_ratio: string;
  verdict: 'caution' | 'not_recommended' | 'safe';
  suggested_action: string;
};

export type PurchaseCheckResponse = {
  verdict: 'caution' | 'not_recommended' | 'safe';
  affordability_score: number;
  guidance: string;
  provider: string;
  model_name: string;
  context: PurchaseCheckContext;
};

export async function purchaseCheck(accessToken: string, payload: PurchaseCheckPayload) {
  return apiRequest<PurchaseCheckResponse>('/ai/purchase-check', {
    accessToken,
    body: payload,
    method: 'POST',
  });
}

export type SavingsAdvicePayload = {
  question: string;
  goal_id?: string | null;
};

export type SavingsAdviceAllocation = {
  goal_id: string;
  name: string;
  recommended_monthly_contribution: string;
  required_monthly_contribution: string;
  pace_status: 'at_risk' | 'behind' | 'on_track';
};

export type SavingsAdviceContext = {
  currency_code: string;
  currency_symbol: string;
  month_label: string;
  current_month_income: string;
  current_month_expense: string;
  current_month_net: string;
  active_goal_count: number;
  comfortable_monthly_savings: string;
  total_goal_monthly_required: string;
  overall_goal_progress: string;
  can_fund_all_goals_on_time: boolean;
  recommendation_text: string;
  focus_goal_name: string | null;
  focus_goal_progress_percentage: string | null;
  focus_goal_monthly_required: string | null;
  focus_goal_pace_status: 'at_risk' | 'behind' | 'on_track' | null;
  allocations: SavingsAdviceAllocation[];
};

export type SavingsAdviceResponse = {
  guidance: string;
  provider: string;
  model_name: string;
  context: SavingsAdviceContext;
};

export async function savingsAdvice(accessToken: string, payload: SavingsAdvicePayload) {
  return apiRequest<SavingsAdviceResponse>('/ai/savings-advice', {
    accessToken,
    body: payload,
    method: 'POST',
  });
}
